src/nfa/operations.cc

```diff
@@ -1207,27 +1205,68 @@ OrdVector<Symbol> mata::nfa::get_symbols_to_work_with(const Nfa& nfa, const mata
 
 std::optional<mata::Word> Nfa::get_word(const Symbol first_epsilon) const {
     if (initial.empty() || final.empty()) { return std::nullopt; }
+    if (initial.intersects_with(final)) { return Word{}; }
 
-    std::vector<std::pair<State, Word>> worklist{};
+    /// Current state state post iterator, its end iterator, and iterator in the current symbol post to target states.
+    std::vector<std::tuple<StatePost::const_iterator, StatePost::const_iterator, StateSet::const_iterator>> worklist{};
+    std::vector<bool> searched(num_of_states(), false);
+    bool final_found{};
     for (const State initial_state: initial) {
         if (final.contains(initial_state)) { return Word{}; }
-        worklist.emplace_back(initial_state, Word{});
-    }
-    std::vector<bool> searched(num_of_states());
+        if (searched[initial_state]) { continue; }
+        searched[initial_state] = true;
+
+        const StatePost& initial_state_post{ delta[initial_state] };
+        auto initial_symbol_post_it{ initial_state_post.cbegin() };
+        auto initial_symbol_post_end{ initial_state_post.cend() };
+        while (worklist.empty() && initial_symbol_post_it != initial_state_post.cend()) {
+            auto initial_target_it{ initial_symbol_post_it->targets.cbegin() };
+            do {
+                if (searched[*initial_target_it]) {
+                    ++initial_target_it;
+                } else {
+                    worklist.emplace_back(initial_symbol_post_it, initial_symbol_post_end, initial_target_it);
+                    break;
+                }
+            } while (initial_target_it != initial_symbol_post_it->targets.cend());
+            if (initial_target_it == initial_symbol_post_it->targets.cend()) {
+                ++initial_symbol_post_it;
+            }
+        }
 
-    while (!worklist.empty()) {
-        auto [state, word]{ std::move(worklist.back()) };
-        worklist.pop_back();
-        for (const Move move: delta[state].moves()) {
-            if (searched[move.target]) { continue; }
-            Word target_word{ word };
-            if (move.symbol < first_epsilon) { target_word.push_back(move.symbol); }
-            if (final.contains(move.target)) { return target_word; }
-            worklist.emplace_back(move.target, target_word);
-            searched[move.target] = true;
+        while (!worklist.empty()) {
+            // Using references to iterators to be able to increment the top-most element in the worklist in place.
+            auto& [state_post_it, state_post_end, targets_it]{ worklist.back() };
+            if (state_post_it != state_post_end) {
+                if (targets_it != state_post_it->targets.cend()) {
+                    if (searched[*targets_it]) { ++targets_it; continue; }
+                    if (final.contains(*targets_it)) { final_found = true; break; }
```

Comment:
This breaks from the `while` loop, but not from the `for` loop

Reply:
Which would mean that we would go to the next initial state.

Reply:
I accidentally removed an if statement breaking from the outer initial state loop when refactoring. Good catch. I also add a new test for this case.

Reply:
PS: It actually worked by accident even with the if statement removed. If the previous initial state found a final state, at least some transitions were left in the worklist, so for the next initial state no successors were added since the worklist was not empty, and we immediately continued to the next initial state.